src/cppnv/EnvReader.cc

```diff
@@ -0,0 +1,813 @@
+  interpolation->closed = true;
+  value->interpolation_index++;
+}
+
+void EnvReader::open_variable(EnvValue* value) {
+  int position;
+  const auto result = position_of_dollar_last_sign(value, &position);
+
+  if (result == success) {
+    value->is_parsing_variable = true;
+    value->interpolations->push_back(
+        new VariablePosition(value->value_index,
+                             value->value_index - 1,
+                             position));
+  }
+}
+
+bool EnvReader::walk_double_quotes(EnvValue* value) {
+  // we have have some quotes at the start
+  if (value->value_index == 0) {
+    if (value->double_quote_streak == 1) {
+      value->double_quote_streak = 0;
+      value->double_quoted = true;
+      return false;  // we have a double quote at the start
+    }
+    // we have a empty double quote value aka ''
+    if (value->double_quote_streak == 2) {
+      value->double_quote_streak = 0;
+      value->double_quoted = true;
+      return true;  // we have a  empty double quote at the start
+    }
+    if (value->double_quote_streak == 3) {
+      value->double_quote_streak = 0;
+      value->triple_double_quoted = true;
+      return false;  // we have a triple double quote at the start
+    }
+    if (value->double_quote_streak > 5) {
+      value->double_quote_streak = 0;
+      value->triple_double_quoted = true;
+      // basically we have """""" an empty heredoc with extra " at the end.
+      // Ignore the trailing "
+      return true;  // we have a triple quote at the start
+    }
+    if (value->double_quote_streak > 3) {
+      value->triple_double_quoted = true;
+      // we have """"...
+      // add the diff to the buffer
+      for (int i = 0; i < value->double_quote_streak - 3; i++) {
+        add_to_buffer(value, '"');
+      }
+      value->double_quote_streak = 0;
+      return false;  // we have a triple quote at the start
+    }
+
+    return false;  // we have garbage
+  }
+
+  // we're single quoted
+  if (value->double_quoted) {
+    // any amount of quotes sends
+    value->double_quote_streak = 0;
+    return true;
+  }
+
+  // We have a triple quote
+  if (value->triple_double_quoted) {
+    if (value->double_quote_streak == 3 || value->double_quote_streak > 3) {
+      value->double_quote_streak = 0;
+      return true;  // we have enough to close, truncate trailing single quotes
+    }
+    // we have not enough to close the heredoc.
+    if (value->double_quote_streak < 3) {
+      // add them to the buffer
+      for (int i = 0; i < value->double_quote_streak; i++) {
+        add_to_buffer(value, '"');
+      }
+      value->double_quote_streak = 0;
+      return false;
+    }
+    return false;  // we have garbage
+  }
+
+  return false;
+}
+
+/**
+ * \brief
+ * \param value
+ * \return True if end quotes detected and input should stop, false otherwise
+ */
+bool EnvReader::walk_single_quotes(EnvValue* value) {
+  // if (value->quoted && value->single_quote_streak == 0) {
```

Comment:
unused method, we can remove this :)

Reply:
I don't think this is unused. In fact, it's one of the most important to single quote heredocs.

Reply:
Sorry I mean if there is commented code, it should not be added to the commit

Reply:
Comments removed.